tfe/data_source_oauth_clients.go

```diff
@@ -0,0 +1,138 @@
+package tfe
+
+import (
+	"fmt"
+	"log"
+	"time"
+
+	tfe "github.com/hashicorp/go-tfe"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+)
+
+func dataSourceTFEOauthClients() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceTFEOauthClientList,
+
+		Schema: map[string]*schema.Schema{
+			"organization": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+			"oauth_clients": {
+				Computed: true,
+				Type:     schema.TypeList,
+				Elem:     &schema.Schema{Type: schema.TypeMap},
+			},
+		},
+	}
+}
+
+func dataSourceTFEOauthClientList(d *schema.ResourceData, meta interface{}) error {
+	tfeClient := meta.(*tfe.Client)
+
+	organization := d.Get("organization").(string)
+
+	log.Printf("[DEBUG] Setting OauthClients Attributes for organization %s", organization)
+	err := oauthClientsPopulateFields(tfeClient, organization, d)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func oauthClientsPopulateFields(client *tfe.Client, organization string, d *schema.ResourceData) error {
+	var oauthClients []map[string]interface{}
+
+	log.Printf("[DEBUG] Listing oauth clients")
+	options := tfe.OAuthClientListOptions{
+		ListOptions: tfe.ListOptions{
+			PageSize: 100,
+		},
+	}
+	for {
+		oauthClientList, err := client.OAuthClients.List(ctx, organization, options)
+		if err != nil {
+			return fmt.Errorf("failed to retrieve oauth Clients from organization %s: %v", organization, err)
+		}
+
+		for _, client := range oauthClientList.Items {
+			flattenedOauthClient, err := flattenOauthClient(client)
+			if err != nil {
+				return fmt.Errorf("failed to flatten oauthClient: %v", err)
+			}
+			log.Printf("[DEBUG] Flattened OauthClient: %v", flattenedOauthClient)
+
+			oauthClients = append(oauthClients, flattenedOauthClient)
+		}
+
+		// Exit the loop when we've seen all pages.
+		if oauthClientList.CurrentPage >= oauthClientList.TotalPages {
+			break
+		}
+
+		// Update the page number to get the next page.
+		options.PageNumber = oauthClientList.NextPage
+	}
+
+	d.SetId("oauth-clients")
+	return d.Set("oauth_clients", oauthClients)
+}
+
+func flattenOauthClient(in *tfe.OAuthClient) (map[string]interface{}, error) {
```

Comment:
I'm curious about this one. Is there a particular reason you wish to flatten the `OAuthClient` object? I'm wondering about the implications if a user expects to fetch an attribute that was omitted due to flattening.

Reply:
Also, this isn't a pattern that I've seen applied to other data sources, so I'm hesitant to introduce this.